cmd/sops/subcommand/publish/publish.go

```diff
@@ -0,0 +1,114 @@
+package publish
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"go.mozilla.org/sops"
+	"go.mozilla.org/sops/cmd/sops/codes"
+	"go.mozilla.org/sops/cmd/sops/common"
+	"go.mozilla.org/sops/config"
+	"go.mozilla.org/sops/keyservice"
+	"go.mozilla.org/sops/logging"
+
+	"github.com/sirupsen/logrus"
+)
+
+var log *logrus.Logger
+
+func init() {
+	log = logging.NewLogger("PUBLISH")
+}
+
+type Opts struct {
+	Interactive bool
+	ConfigPath  string
+	InputPath   string
+	KeyServices []keyservice.KeyServiceClient
+	InputStore  sops.Store
+}
+
+func Run(opts Opts) error {
+	var fileContents []byte
+	path, err := filepath.Abs(opts.InputPath)
+	if err != nil {
+		return err
+	}
+	info, err := os.Stat(path)
+	if err != nil {
+		return err
+	}
+	if info.IsDir() {
+		return fmt.Errorf("can't operate on a directory")
+	}
+	_, fileName := filepath.Split(path)
+
+	conf, err := config.LoadForFile(opts.ConfigPath, opts.InputPath, make(map[string]*string))
+	if err != nil {
+		return err
+	}
+	if conf.Destination == nil {
+		return errors.New("No destination configured for this file.")
+	}
+
+	// Check that this is a sops-encrypted file
+	tree, err := common.LoadEncryptedFile(opts.InputStore, opts.InputPath)
+	if err != nil {
+		return err
+	}
+
+	// Re-encrypt if settings exist to do so
+	if len(conf.ReEncryptionKeyGroups) != 0 {
+		key, err := tree.Metadata.GetDataKeyWithKeyServices(opts.KeyServices)
```

Comment:
should this be interactive to make sure the user accepts to remove the existing keygroups and reencrypt?

Reply:
It might be worth adding the confirmation step prior to this and include it within the confirmation message (similar to `sops updatekeys`). ie:
```
$ sops publish s3/app.yaml
uploading /home/ajvb/sops-testing/publish/s3/app.yaml to s3://sops-secrets/testing/app.yaml

and the following changes will be made to the file's groups:
+++ F69E4901EDBAD2D1753F8C67A64535C4163FB307
--- 1022470DE3F0BC54BC6AB62DE05550BC07FB1A0A

Is this okay? (y/n): 
```